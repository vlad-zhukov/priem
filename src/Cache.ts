const NEXT = '@@next';
const PREV = '@@prev';

export class CacheItem<K, V> {
    key: K;
    value: V;
    [NEXT]: CacheItem<K, V> | undefined;
    [PREV]: CacheItem<K, V> | undefined;
    isValid!: boolean;
    expireTimerId: number | undefined;
    lastUpdateAt: number | undefined;

    constructor(key: K, value: V) {
        this.key = key;
        this.value = value;

        Object.defineProperties(this, {
            [NEXT]: {
                value: undefined,
                writable: true,
            },
            [PREV]: {
                value: undefined,
                writable: true,
            },
            isValid: {
                value: false,
                writable: true,
            },
            expireTimerId: {
                value: undefined,
                writable: true,
            },
            lastUpdateAt: {
                value: undefined,
                writable: true,
            },
        });
    }

    destroy(): void {
        delete this.key;
        delete this.value;
        this[NEXT] = undefined;
        this[PREV] = undefined;
        this.isValid = false;
        clearTimeout(this.expireTimerId);
        this.expireTimerId = undefined;
        this.lastUpdateAt = undefined;
    }
}

export interface SerializableCacheItem<K, V> {
    key: K;
    value: V;
}

const REDUCED = '@@reduced';

interface ReducedType<ValueType> {
    [REDUCED]: true;
    value: ValueType;
}

function reduced<ValueType>(value: ValueType): ReducedType<ValueType> {
    return {[REDUCED]: true, value};
}

function isReduced<ValueType>(value: any): value is ReducedType<ValueType> {
    return value && REDUCED in value;
}

export function reduce<ReturnType, K, V>(
    cache: Cache<K, V>,
    accumulator: ReturnType,
    iteratee: (result: ReturnType, item: CacheItem<K, V>) => ReturnType | ReducedType<ReturnType>,
) {
    let item = cache.head;
    let result: ReturnType | ReducedType<ReturnType> = accumulator;
    while (item) {
        result = iteratee(result, item);
        if (isReduced(result)) {
            return result.value;
        }
        item = item[NEXT];
    }
    return result;
}

export class Cache<K, V> {
    head: CacheItem<K, V> | undefined;
    tail: CacheItem<K, V> | undefined;
    size = 0;

    constructor(items: (CacheItem<K, V> | SerializableCacheItem<K, V>)[]) {
        const now = Date.now();

        for (let i = items.length; i > 0; i--) {
            const item = items[i - 1];
            const cacheItem: CacheItem<K, V> =
                item instanceof CacheItem ? item : new CacheItem<K, V>(item.key, item.value);
            cacheItem.isValid = true;
            cacheItem.lastUpdateAt = now;
            this.prepend(cacheItem);
        }
    }

    prepend(item: CacheItem<K, V>): void {
        item[NEXT] = this.head;
        if (this.head) {
            this.head[PREV] = item;
        }
        this.head = item;
        if (!this.tail) {
            this.tail = item;
        }
        this.size += 1;
    }

    remove(item: CacheItem<K, V>): boolean {
        if (!item.key) {
            return false;
        }

        const next = item[NEXT];
        const prev = item[PREV];

        if (prev) {
            prev[NEXT] = next;
        } else {
            this.head = next;
        }

        if (next) {
            next[PREV] = prev;
        } else {
            this.tail = prev;
        }

        this.size -= 1;

        return true;
    }

    findBy(predicate: (item: CacheItem<K, V>) => boolean): CacheItem<K, V> | undefined {
        return reduce<CacheItem<K, V> | undefined, K, V>(this, undefined, (acc, item) =>
            predicate(item) ? reduced(item) : acc,
        );
    }
}
