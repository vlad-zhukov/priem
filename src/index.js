export {default as Priem} from './Priem';
export {Container, populateStore, flushStore} from './Container';
export {default as withPriem} from './withPriem';
export {default as getDataFromTree} from './getDataFromTree';
