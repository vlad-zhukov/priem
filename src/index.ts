export {createResource, GetArgs, Options, Result, ResultMeta, ResultPages, ResultPagesMeta} from './createResource';
export {
    hydrateStore,
    flushStore,
    getRunningPromises,
    ResourceOptions,
    MemoizedKey,
    MemoizedValue,
    MemoizedSerializableCacheItem,
    Status,
} from './Resource';
export {SerializableCacheItem} from './Cache';
