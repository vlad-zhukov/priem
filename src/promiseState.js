import {type} from './helpers';

/**
 * Constructor for creating a new promiseState.
 * @param [options] {Object}
 * @param [options.pending] {Boolean}
 * @param [options.refreshing] {Boolean}
 * @param [options.fulfilled] {Boolean}
 * @param [options.rejected] {Boolean}
 * @param [options.value] {*}
 * @param [options.reason] {String|Null}
 * @returns {Object}
 */
function create({
    pending = false, // eslint-disable-line no-shadow
    refreshing = false, // eslint-disable-line no-shadow
    fulfilled = false, // eslint-disable-line no-shadow
    rejected = false, // eslint-disable-line no-shadow
    value = null,
    reason = null,
}) {
    return {
        pending,
        refreshing,
        fulfilled,
        rejected,
        value,
        reason,
    };
}

/**
 * Checks if a passed value is a promiseState-like object.
 * @param maybePromiseState {*}
 * @returns {Boolean}
 */
export function isPromiseState(maybePromiseState) {
    if (type(maybePromiseState) !== 'object') {
        return false;
    }

    const propertyNames = ['pending', 'refreshing', 'fulfilled', 'rejected'];
    return propertyNames.every(propertyName => type(maybePromiseState[propertyName]) === 'boolean');
}

/**
 * Checks if a passed promiseState is in a loading state.
 * @param promiseState {*}
 * @returns {Boolean}
 */
export function isLoading(promiseState) {
    if (isPromiseState(promiseState) === false) {
        throw new TypeError('Priem: Expected a promiseState.');
    }
    return promiseState.pending || promiseState.refreshing;
}

/**
 * Creates a new promiseState that is pending.
 * @returns {Object}
 */
export function pending() {
    return create({pending: true});
}

/**
 * Creates a promiseState that is refreshing.
 * Can be called without a previous promiseState and will be both pending and refreshing.
 * @param [previous] {Object}
 * @returns {Object}
 */
export function refreshing(previous) {
    const previousPromiseState = isPromiseState(previous) ? previous : pending();
    return create({...previousPromiseState, refreshing: true});
}

/**
 * Creates a promiseState that is resolved with the given value.
 * If the given value is already a promiseState, its `value` property will be used instead.
 * @param [valueOrPromiseState] {*}
 * @returns {Object}
 */
export function fulfilled(valueOrPromiseState) {
    const value = isPromiseState(valueOrPromiseState) ? valueOrPromiseState.value : valueOrPromiseState;
    return create({fulfilled: true, value});
}

/**
 * Creates a promiseState that is rejected with the given reason.
 * @param reason {String}
 * @returns {Object}
 */
export function rejected(reason) {
    return create({rejected: true, reason});
}
