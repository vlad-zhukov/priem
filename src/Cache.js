import moize from 'moize';
import {elementsEqual} from 'react-shallow-equal';
import * as promiseState from './promiseState';
import {type, assertType, isBrowser} from './helpers';

function arrayElementsEqual(a, b) {
    if (!Array.isArray(a) || !Array.isArray(b) || a.length !== b.length) {
        return false;
    }
    for (let i = 0; i < a.length; i++) {
        if (a[i] !== b[i]) {
            return false;
        }
    }

    return true;
}

export const defaultAsyncValueOptions = {autoRefresh: true, maxSize: Infinity};

export default class Cache {
    constructor() {
        this.memoized = null;
        this.awaiting = [];
        this.rejected = false;
        this.prevArgs = null;
    }

    add(value, onExpire) {
        assertType(value, ['object'], "'getAsyncValue() -> value'");
        assertType(value.promise, ['function'], "'getAsyncValue() -> value.promise'");

        this.memoized = moize(value.promise, {
            isPromise: true,
            maxAge: value.maxAge,
            maxArgs: value.maxArgs,
            maxSize: value.maxSize,
            onExpire,
        });
    }

    isMemoized(args) {
        if (!this.memoized) {
            return false;
        }

        return this.memoized.has(args);
    }

    getAwaitingIndex(args) {
        for (let i = 0, l = this.awaiting.length; i < l; i++) {
            if (elementsEqual(this.awaiting[i], args)) {
                return i;
            }
        }

        return -1;
    }

    isAwaiting(args) {
        return this.getAwaitingIndex(args) !== -1;
    }

    removeAwaiting(args) {
        const index = this.getAwaitingIndex(args);
        if (index !== -1) {
            this.awaiting.splice(index, 1);
        }
    }

    run({asyncValue, isForced, update, onExpire}) {
        const value = {...defaultAsyncValueOptions, ...asyncValue};

        // Stop auto-refreshing
        if (!value.autoRefresh && !isForced) {
            return;
        }

        const args = type(value.args) === 'array' ? value.args : [];

        if (!this.memoized) {
            this.add(value, onExpire);
        }

        // Do not recall rejected (uncached) promises unless forced
        if (this.rejected && !isForced) {
            return;
        }

        if (this.isAwaiting(args)) {
            return;
        }

        let shouldUpdateState = true;

        // eslint-disable-next-line consistent-return
        update((s, m) => {
            const isFulfilled = !!s.fulfilled;
            const isSsr = !!(isFulfilled && m.ssr);

            if (this.isMemoized(args)) {
                // Do not recall memoized promises unless forced
                if (isFulfilled && !isForced && arrayElementsEqual(args, this.prevArgs)) {
                    shouldUpdateState = false;
                    return null;
                }
            }
            else if (isSsr) {
                shouldUpdateState = false;
                this.memoized.add(args, Promise.resolve(s.value));
                this.prevArgs = args;
                // eslint-disable-next-line consistent-return
                return {
                    meta: {ssr: false},
                };
            }

            this.awaiting.unshift(args);
            this.rejected = false;

            if (isFulfilled) {
                return {
                    state: promiseState.refreshing(s),
                };
            }

            return {
                state: promiseState.pending(),
            };
        });

        if (!shouldUpdateState) {
            return;
        }

        if (isForced) {
            this.memoized.remove(args);
        }

        // eslint-disable-next-line consistent-return
        return this.memoized(...args)
            .then((result) => {
                this.removeAwaiting(args);
                this.prevArgs = args;
                update({
                    state: promiseState.fulfilled(result),
                    meta: {ssr: !isBrowser},
                });
            })
            .catch((e) => {
                this.removeAwaiting(args);
                this.rejected = true;
                this.prevArgs = null;
                update({
                    state: promiseState.rejected(e.message),
                    meta: {ssr: !isBrowser},
                });
            });
    }
}
