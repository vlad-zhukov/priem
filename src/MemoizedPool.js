import moize from 'moize';
import {elementsEqual} from 'react-shallow-equal';
import * as promiseState from './promiseState';
import {type, isBrowser} from './helpers';

function arrayElementsEqual(a, b) {
    if (!Array.isArray(a) || !Array.isArray(b) || a.length !== b.length) {
        return false;
    }
    for (let i = 0; i < a.length; i++) {
        if (a[i] !== b[i]) {
            return false;
        }
    }

    return true;
}

export const defaultAsyncValueOptions = {autoRefresh: true, maxSize: Infinity};

export function extractAsyncValues(props) {
    if (type(props.asyncValues) !== 'function') {
        return {};
    }

    const values = props.asyncValues(props);

    const typeOfValues = type(values);
    if (typeOfValues !== 'object') {
        throw new TypeError(
            `Priem: property 'asyncValues' must be a function that returns an object, but got: '${typeOfValues}'.`
        );
    }

    return values;
}

export class MemoizedPool {
    constructor() {
        this.memoized = {};
        this.awaiting = {};
        this.rejected = {};
    }

    addMemoized(key, value, onExpire) {
        const typeOfValue = type(value);
        if (typeOfValue !== 'object') {
            throw new TypeError(
                `Priem: argument 'values' must return an object of objects, but got: '${typeOfValue}'.`
            );
        }

        const typeOfValuePromise = type(value.promise);
        if (typeOfValuePromise !== 'function') {
            throw new TypeError(
                "Priem: argument 'values' must return an object of objects with a property 'promise' " +
                    `each, but got: '${typeOfValuePromise}'.`
            );
        }

        this.awaiting[key] = [];
        this.rejected[key] = false;
        this.memoized[key] = moize(value.promise, {
            isPromise: true,
            maxAge: value.maxAge,
            maxArgs: value.maxArgs,
            maxSize: value.maxSize,
            onExpire,
        });
    }

    isMemoized(key, args) {
        if (!this.memoized[key]) {
            return false;
        }

        return this.memoized[key].has(args);
    }

    getAwaitingIndex(key, args) {
        for (let i = 0, l = this.awaiting[key].length; i < l; i++) {
            if (elementsEqual(this.awaiting[key][i], args)) {
                return i;
            }
        }

        return -1;
    }

    isAwaiting(key, args) {
        return this.getAwaitingIndex(key, args) !== -1;
    }

    removeAwaiting(key, args) {
        const index = this.getAwaitingIndex(key, args);
        if (index !== -1) {
            this.awaiting[key].splice(index, 1);
        }
    }

    runPromise({key, value, publicKey, isForced, update, onExpire}) {
        const args = type(value.args) === 'array' ? value.args : [];

        if (!this.memoized[key]) {
            this.addMemoized(key, value, onExpire);
        }

        // Do not recall rejected (uncached) promises unless forced
        if (this.rejected[key] && !isForced) {
            return;
        }

        if (this.isAwaiting(key, args)) {
            return;
        }

        // eslint-disable-next-line consistent-return
        return update((s, m) => {
            const isFulfilled = !!s?.[publicKey]?.fulfilled; // eslint-disable-line no-undef
            const isSsr = !!(isFulfilled && m?.[publicKey]?.ssr); // eslint-disable-line no-undef

            if (this.isMemoized(key, args)) {
                // Do not recall memoized promises unless forced
                if (isFulfilled && !isForced && arrayElementsEqual(args, m?.[publicKey]?.args)) {
                    return null;
                }
            }
            else if (isSsr) {
                this.memoized[key].add(args, Promise.resolve(s[publicKey].value));
                // eslint-disable-next-line consistent-return
                return {
                    meta: {[publicKey]: {ssr: false}},
                };
            }

            this.awaiting[key].unshift(args);
            this.rejected[key] = false;

            if (isFulfilled) {
                return {
                    data: {
                        [publicKey]: promiseState.refreshing(s[publicKey]),
                    },
                };
            }

            return {
                data: {
                    [publicKey]: promiseState.pending(),
                },
            };
        }).then((s) => {
            if (!s) {
                return null;
            }

            return this.memoized[key](...args)
                .then((result) => {
                    this.removeAwaiting(key, args);
                    return update({
                        data: {[publicKey]: promiseState.fulfilled(result)},
                        meta: {[publicKey]: {ssr: !isBrowser, args}},
                    });
                })
                .catch((e) => {
                    this.removeAwaiting(key, args);
                    this.rejected[key] = true;
                    return update({
                        data: {[publicKey]: promiseState.rejected(e.message)},
                        meta: {[publicKey]: {ssr: !isBrowser, args: null}},
                    });
                });
        });
    }

    runPromises({props, isForced, update, onExpire}) {
        const asyncValues = extractAsyncValues(props);

        const values = Object.keys(asyncValues).map((key) => {
            const asyncValue = {...defaultAsyncValueOptions, ...asyncValues[key]};

            // Stop auto-refreshing
            if (!asyncValue.autoRefresh && !isForced) {
                return null;
            }

            return this.runPromise({
                key: `${key}@${props.name}`,
                value: asyncValue,
                publicKey: key,
                isForced,
                update,
                onExpire,
            });
        });

        // eslint-disable-next-line consistent-return
        return Promise.all(values);
    }
}
