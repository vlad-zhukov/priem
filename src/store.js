import * as promiseState from './promiseState';
import {extractAsyncValues} from './MemoizedPool';
import {type, isBrowser} from './helpers';

export const consts = {
    NAME: '@@name',
    INITIAL_VALUES: '@@initialValues',
    COUNT: '@@count',
    PERSIST: '@@persist',
};

export function createInitializeFunction(component, isFake = false) {
    const countStep = isFake ? 0 : 1;

    return props =>
        new Promise((resolve) => {
            component.setState(
                (state) => {
                    const s = state[props.name];
                    const m = component.meta[props.name];

                    if (s && m) {
                        // eslint-disable-next-line no-param-reassign
                        component.meta[props.name] = {...m, [consts.COUNT]: (m[consts.COUNT] += countStep)};
                        return null;
                    }

                    const initialValues = {...props.priem};
                    const meta = {};

                    if (props.autoRefresh !== false) {
                        const asyncKeys = Object.keys(extractAsyncValues(props));
                        for (let i = 0, l = asyncKeys.length; i < l; i++) {
                            initialValues[asyncKeys[i]] = promiseState.pending();
                            meta[asyncKeys[i]] = {ssr: !isBrowser};
                        }
                    }

                    // eslint-disable-next-line no-param-reassign
                    component.meta[props.name] = {
                        [consts.NAME]: props.name,
                        [consts.INITIAL_VALUES]: props.initialValues,
                        [consts.PERSIST]: props.persist,
                        [consts.COUNT]: countStep,
                        ...meta,
                    };
                    return {
                        [props.name]: initialValues,
                    };
                },
                () => resolve(component.state[props.name])
            );
        });
}

export function createDestroyFunction(component, isFake = false) {
    return name =>
        new Promise((resolve) => {
            if (isFake) {
                resolve(component.state[name]);
                return;
            }

            component.setState(
                (state) => {
                    const s = state[name];
                    const m = component.meta[name];
                    if (s && m && m[consts.COUNT] > 0 && m[consts.PERSIST] === true) {
                        // eslint-disable-next-line no-param-reassign
                        component.meta[name] = {...m, [consts.COUNT]: m[consts.COUNT] - 1};
                        return null;
                    }

                    // eslint-disable-next-line no-param-reassign
                    component.meta[name] = undefined;
                    return {[name]: undefined};
                },
                () => resolve(component.state[name])
            );
        });
}

export function createUpdateFunction(component) {
    return (name, updater) =>
        new Promise((resolve) => {
            let didUpdate = false;
            component.setState(
                (state) => {
                    const s = state[name];
                    const m = component.meta[name];

                    const updaterResult = type(updater) === 'function' ? updater(s, m) : updater;

                    if (updaterResult != null) {
                        // eslint-disable-next-line no-param-reassign
                        component.meta[name] = {...m, ...updaterResult.meta};

                        if (updaterResult.data) {
                            didUpdate = true;
                            return {
                                [name]: {...s, ...updaterResult.data},
                            };
                        }
                    }

                    return null;
                },
                () => resolve(didUpdate ? component.state[name] : undefined)
            );
        });
}

export class FakeProviderStore {
    constructor() {
        this.state = {};
        this.meta = {};

        this.initialize = createInitializeFunction(this, true);
        this.destroy = createDestroyFunction(this, true);
        this.update = createUpdateFunction(this);
    }

    setState(updater, callback) {
        const nextState = type(updater) === 'function' ? updater(this.state) : updater;
        if (nextState != null) {
            this.state = {...this.state, ...nextState};
        }
        if (type(callback) === 'function') {
            callback();
        }
    }
}
