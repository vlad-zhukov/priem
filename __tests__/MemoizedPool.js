import delay from 'delay';
import {MemoizedPool} from '../src/MemoizedPool';
import {FakeProviderStore} from '../src/store';
import * as promiseState from '../src/promiseState';

function setup(options) {
    const store = new FakeProviderStore();
    const pool = new MemoizedPool();

    store.initialize(options.props);

    const update = jest.fn(updater => store.update(options.props.name, updater));
    const onExpire = jest.fn(() => {
        // eslint-disable-next-line no-use-before-define
        runPromises();
    });

    function runPromises(moreOptions) {
        return pool.runPromises({...options, ...moreOptions, onExpire, update});
    }

    return {
        store,
        pool,
        update,
        onExpire,
        runPromises,
    };
}

it('should not run promises if both `autoRefresh` and `isForced` are false', async () => {
    const props = {
        name: 'Test',
        asyncValues: () => ({
            testValue: {
                args: ['foo'],
                promise: value => delay(200, {value}),
                autoRefresh: false,
            },
        }),
    };

    const {store, pool, update, onExpire, runPromises} = setup({props, isForced: false});

    runPromises();
    expect(pool.awaiting).toMatchObject({});

    await delay(250);
    expect(pool.awaiting).toMatchObject({});

    expect(update).toHaveBeenCalledTimes(0);
    expect(onExpire).toHaveBeenCalledTimes(0);
    expect([store.state, store.meta]).toMatchSnapshot();
});

it('should run promises if `autoRefresh` is false but `isForced` is true', async () => {
    const props = {
        name: 'Test',
        asyncValues: () => ({
            testValue: {
                args: ['foo'],
                promise: value => delay(200, {value}),
                autoRefresh: false,
            },
        }),
    };

    const {store, pool, update, onExpire, runPromises} = setup({props, isForced: true});

    runPromises();
    expect(pool.awaiting).toMatchObject({'testValue@Test': [['foo']]});

    await delay(250);
    expect(pool.awaiting).toMatchObject({'testValue@Test': []});

    expect(update).toHaveBeenCalledTimes(2);
    expect(onExpire).toHaveBeenCalledTimes(0);
    expect([store.state, store.meta]).toMatchSnapshot();
});

it('should not run promises when awaiting', async () => {
    const props = {
        name: 'Test',
        asyncValues: () => ({
            testValue: {
                args: ['foo'],
                promise: value => delay(200, {value}),
            },
        }),
    };

    const {store, pool, update, onExpire, runPromises} = setup({props, isForced: false});

    runPromises();
    expect(pool.awaiting).toMatchObject({'testValue@Test': [['foo']]});

    runPromises();
    expect(pool.awaiting).toMatchObject({'testValue@Test': [['foo']]});

    runPromises();
    expect(pool.awaiting).toMatchObject({'testValue@Test': [['foo']]});

    await delay(250);
    expect(pool.awaiting).toMatchObject({'testValue@Test': []});

    expect(update).toHaveBeenCalledTimes(2);
    expect(onExpire).toHaveBeenCalledTimes(0);
    expect([store.state, store.meta]).toMatchSnapshot();
});

it('should refresh promise if forced', async () => {
    const props = {
        name: 'Test',
        asyncValues: () => ({
            testValue: {
                args: ['foo'],
                promise: value => delay(200, {value}),
            },
        }),
    };

    const {store, update, runPromises} = setup({props, isForced: false});

    await runPromises();
    expect([store.state, store.meta]).toMatchSnapshot();

    runPromises({isForced: true});
    expect([store.state, store.meta]).toMatchSnapshot();
    await delay(250);

    expect(update).toHaveBeenCalledTimes(4);
});

it('should not rerun promises if previous promise was rejected', async () => {
    let called = false;
    const props = {
        name: 'Test',
        asyncValues: () => ({
            testValue: {
                args: ['foo'],
                promise(value) {
                    if (!called) {
                        called = true;
                        return delay.reject(200, new Error('foo'));
                    }
                    return delay(200, {value});
                },
            },
        }),
    };

    const {store, pool, update, onExpire, runPromises} = setup({props, isForced: false});

    runPromises();
    expect(pool.awaiting).toMatchObject({'testValue@Test': [['foo']]});

    await delay(250);
    expect(pool.awaiting).toMatchObject({'testValue@Test': []});

    runPromises();
    expect(pool.awaiting).toMatchObject({'testValue@Test': []});

    expect(update).toHaveBeenCalledTimes(2);
    expect(onExpire).toHaveBeenCalledTimes(0);
    expect([store.state, store.meta]).toMatchSnapshot();
});

it('should rerun promises if previous promise was rejected but `isForced` is true', async () => {
    let called = false;
    const props = {
        name: 'Test',
        asyncValues: () => ({
            testValue: {
                args: ['foo'],
                promise(value) {
                    if (!called) {
                        called = true;
                        return delay.reject(200, new Error('foo'));
                    }
                    return delay(200, {value});
                },
            },
        }),
    };

    const {store, pool, update, onExpire, runPromises} = setup({props, isForced: true});

    runPromises();
    expect(pool.awaiting).toMatchObject({'testValue@Test': [['foo']]});

    await delay(250);
    expect(pool.awaiting).toMatchObject({'testValue@Test': []});

    runPromises();
    expect(pool.awaiting).toMatchObject({'testValue@Test': [['foo']]});

    await delay(250);
    expect(pool.awaiting).toMatchObject({'testValue@Test': []});

    expect(update).toHaveBeenCalledTimes(4);
    expect(onExpire).toHaveBeenCalledTimes(0);
    expect([store.state, store.meta]).toMatchSnapshot();
});

it('should not try to rerun fulfilled promises if `value` is null', async () => {
    const props = {
        name: 'Test',
        asyncValues: () => ({
            testValue: {
                args: [null],
                promise: value => delay(100, value),
            },
        }),
    };

    const {store, pool, update, onExpire, runPromises} = setup({props, isForced: false});

    runPromises();
    expect(pool.awaiting).toMatchObject({'testValue@Test': [[null]]});

    await delay(150);
    expect(pool.awaiting).toMatchObject({'testValue@Test': []});

    runPromises();
    expect(pool.awaiting).toMatchObject({'testValue@Test': []});

    expect(update).toHaveBeenCalledTimes(3);
    expect(onExpire).toHaveBeenCalledTimes(0);
    expect([store.state, store.meta]).toMatchSnapshot();
});

it('should expire if maxAge is set', async () => {
    const props = {
        name: 'Test',
        asyncValues: () => ({
            testValue: {
                args: ['foo'],
                promise: value => delay(200, {value}),
                maxAge: 300,
            },
        }),
    };

    const {store, pool, update, onExpire, runPromises} = setup({props, isForced: false});

    runPromises();
    expect(pool.awaiting).toMatchObject({'testValue@Test': [['foo']]});

    await delay(250);
    expect([store.state, store.meta]).toMatchSnapshot(); // fulfilled

    await delay(100);
    expect([store.state, store.meta]).toMatchSnapshot(); // refreshing

    await delay(650);
    expect([store.state, store.meta]).toMatchSnapshot(); // refreshing
    expect(pool.awaiting).toMatchObject({'testValue@Test': [['foo']]});

    expect(update).toHaveBeenCalledTimes(7);
    expect(onExpire).toHaveBeenCalledTimes(3);
});

it('should rehydrate ssr data', async () => {
    const promiseFn = jest.fn(value => delay(200, {value}));

    const props = {
        name: 'Test',
        asyncValues: () => ({
            testValue: {
                args: ['foo'],
                promise: promiseFn,
            },
        }),
    };

    const {store, pool, update, onExpire, runPromises} = setup({props, isForced: false});

    await store.update('Test', {
        data: {testValue: promiseState.fulfilled('foo')},
        meta: {testValue: {ssr: true}},
    });

    await runPromises();

    expect([store.state, store.meta]).toMatchSnapshot();

    expect(await pool.memoized['testValue@Test'].get(['foo'])).toBe('foo');

    expect(promiseFn).toHaveBeenCalledTimes(0);
    expect(update).toHaveBeenCalledTimes(1);
    expect(onExpire).toHaveBeenCalledTimes(0);
});

it('should add values to cache when `args` change', async () => {
    let id = 0;
    const props = {
        name: 'Test',
        asyncValues: () => ({
            testValue: {
                args: [`foo${id}`, `bar${id}`],
                promise(foo, bar) {
                    id += 1;
                    return delay(200, {foo, bar});
                },
            },
        }),
    };

    const {pool, update, runPromises} = setup({props, isForced: false});

    await runPromises();
    await runPromises();

    expect(await pool.memoized['testValue@Test'].keys()).toEqual([['foo1', 'bar1'], ['foo0', 'bar0']]);

    expect(update).toHaveBeenCalledTimes(4);
});

it('should return cached values when `args` change', async () => {
    let check = false;
    const props = {
        name: 'Test',
        asyncValues: () => ({
            testValue: {
                args: [`foo-${check}`],
                promise(value) {
                    check = !check;
                    return delay(200, {value});
                },
            },
        }),
    };

    const {store, runPromises} = setup({props, isForced: false});

    await runPromises();
    expect([store.state, store.meta]).toMatchSnapshot(); // foo-false

    await runPromises();
    expect([store.state, store.meta]).toMatchSnapshot(); // foo-true

    await runPromises();
    expect([store.state, store.meta]).toMatchSnapshot(); // foo-false

});
