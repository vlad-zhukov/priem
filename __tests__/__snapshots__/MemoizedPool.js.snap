// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should expire if maxAge is set 1`] = `
Array [
  Object {
    "Test": Object {
      "testValue": Object {
        "fulfilled": true,
        "pending": false,
        "reason": null,
        "refreshing": false,
        "rejected": false,
        "value": Object {
          "value": "foo",
        },
      },
    },
  },
  Object {
    "Test": Object {
      "@@count": 0,
      "@@initialValues": undefined,
      "@@name": "Test",
      "@@persist": undefined,
      "testValue": Object {
        "ssr": false,
      },
    },
  },
]
`;

exports[`should expire if maxAge is set 2`] = `
Array [
  Object {
    "Test": Object {
      "testValue": Object {
        "fulfilled": true,
        "pending": false,
        "reason": null,
        "refreshing": true,
        "rejected": false,
        "value": Object {
          "value": "foo",
        },
      },
    },
  },
  Object {
    "Test": Object {
      "@@count": 0,
      "@@initialValues": undefined,
      "@@name": "Test",
      "@@persist": undefined,
      "testValue": Object {
        "ssr": false,
      },
    },
  },
]
`;

exports[`should expire if maxAge is set 3`] = `
Array [
  Object {
    "Test": Object {
      "testValue": Object {
        "fulfilled": true,
        "pending": false,
        "reason": null,
        "refreshing": true,
        "rejected": false,
        "value": Object {
          "value": "foo",
        },
      },
    },
  },
  Object {
    "Test": Object {
      "@@count": 0,
      "@@initialValues": undefined,
      "@@name": "Test",
      "@@persist": undefined,
      "testValue": Object {
        "ssr": false,
      },
    },
  },
]
`;

exports[`should not rerun promises if previous promise was rejected 1`] = `
Array [
  Object {
    "Test": Object {
      "testValue": Object {
        "fulfilled": false,
        "pending": false,
        "reason": "foo",
        "refreshing": false,
        "rejected": true,
        "value": null,
      },
    },
  },
  Object {
    "Test": Object {
      "@@count": 0,
      "@@initialValues": undefined,
      "@@name": "Test",
      "@@persist": undefined,
      "testValue": Object {
        "ssr": false,
      },
    },
  },
]
`;

exports[`should not run promises if both \`autoRefresh\` and \`isForced\` are false 1`] = `
Array [
  Object {
    "Test": Object {},
  },
  Object {
    "Test": Object {
      "@@count": 0,
      "@@initialValues": undefined,
      "@@name": "Test",
      "@@persist": undefined,
    },
  },
]
`;

exports[`should not run promises when awaiting 1`] = `
Array [
  Object {
    "Test": Object {
      "testValue": Object {
        "fulfilled": true,
        "pending": false,
        "reason": null,
        "refreshing": false,
        "rejected": false,
        "value": Object {
          "value": "foo",
        },
      },
    },
  },
  Object {
    "Test": Object {
      "@@count": 0,
      "@@initialValues": undefined,
      "@@name": "Test",
      "@@persist": undefined,
      "testValue": Object {
        "ssr": false,
      },
    },
  },
]
`;

exports[`should not try to rerun fulfilled promises if \`value\` is null 1`] = `
Array [
  Object {
    "Test": Object {
      "testValue": Object {
        "fulfilled": true,
        "pending": false,
        "reason": null,
        "refreshing": false,
        "rejected": false,
        "value": null,
      },
    },
  },
  Object {
    "Test": Object {
      "@@count": 0,
      "@@initialValues": undefined,
      "@@name": "Test",
      "@@persist": undefined,
      "testValue": Object {
        "ssr": false,
      },
    },
  },
]
`;

exports[`should refresh promise if forced 1`] = `
Array [
  Object {
    "Test": Object {
      "testValue": Object {
        "fulfilled": true,
        "pending": false,
        "reason": null,
        "refreshing": false,
        "rejected": false,
        "value": Object {
          "value": "foo",
        },
      },
    },
  },
  Object {
    "Test": Object {
      "@@count": 0,
      "@@initialValues": undefined,
      "@@name": "Test",
      "@@persist": undefined,
      "testValue": Object {
        "ssr": false,
      },
    },
  },
]
`;

exports[`should refresh promise if forced 2`] = `
Array [
  Object {
    "Test": Object {
      "testValue": Object {
        "fulfilled": true,
        "pending": false,
        "reason": null,
        "refreshing": true,
        "rejected": false,
        "value": Object {
          "value": "foo",
        },
      },
    },
  },
  Object {
    "Test": Object {
      "@@count": 0,
      "@@initialValues": undefined,
      "@@name": "Test",
      "@@persist": undefined,
      "testValue": Object {
        "ssr": false,
      },
    },
  },
]
`;

exports[`should rehydrate ssr data 1`] = `
Array [
  Object {
    "Test": Object {
      "testValue": Object {
        "fulfilled": true,
        "pending": false,
        "reason": null,
        "refreshing": false,
        "rejected": false,
        "value": "foo",
      },
    },
  },
  Object {
    "Test": Object {
      "@@count": 0,
      "@@initialValues": undefined,
      "@@name": "Test",
      "@@persist": undefined,
      "testValue": Object {
        "ssr": false,
      },
    },
  },
]
`;

exports[`should rerun promises if previous promise was rejected but \`isForced\` is true 1`] = `
Array [
  Object {
    "Test": Object {
      "testValue": Object {
        "fulfilled": true,
        "pending": false,
        "reason": null,
        "refreshing": false,
        "rejected": false,
        "value": Object {
          "value": "foo",
        },
      },
    },
  },
  Object {
    "Test": Object {
      "@@count": 0,
      "@@initialValues": undefined,
      "@@name": "Test",
      "@@persist": undefined,
      "testValue": Object {
        "ssr": false,
      },
    },
  },
]
`;

exports[`should run promises if \`autoRefresh\` is false but \`isForced\` is true 1`] = `
Array [
  Object {
    "Test": Object {
      "testValue": Object {
        "fulfilled": true,
        "pending": false,
        "reason": null,
        "refreshing": false,
        "rejected": false,
        "value": Object {
          "value": "foo",
        },
      },
    },
  },
  Object {
    "Test": Object {
      "@@count": 0,
      "@@initialValues": undefined,
      "@@name": "Test",
      "@@persist": undefined,
      "testValue": Object {
        "ssr": false,
      },
    },
  },
]
`;
